text/trace/115-add-sample-to-span-end.md

```diff
@@ -0,0 +1,182 @@
+specific features such as exception handling or deferred close routines.
+
+Deferred span creation has not had as much development focus, which results in
+a less error handling that ensures that deferred spans are being created. This
+results in exceptional situations that occur before the span creation (e.g.
+invalid parsing of an http request) to not be represented in a span, as the
+deferred span is never created.
+
+A way to mitigate this would be to ensure the same level of error handling
+for deferred span creation as with spans themselves (create the span during a
+handled exception). But that would double the amount of maintenance to ensure
+robust span creation. Without span sampling as a concern, it would be desirable
+to just use a span instead, and focus on improving the robustness span creation.
+
+There is another side effect of deferred span creation, where there may be
+child spans that are created before the deferred span is. Examples here include
+a span to represent the receiving of data: the span that represents receiving data
+would have started before the name of the span is available, since that depends
+on receiving and parsing the incoming payload. The result will at worst be
+multiple root spans that should actually be children of the deferred span, since
+there will be no active span present when the children are created.
+
+### It is better to address the issues with sampling
+
+Given the need for an almost identical, most likely less robust code path
+for deferred spans, and the fact that it is a very common pattern, solving
+issues with sampling would ensure a more robust and maintainable code base.
+
+## Explanation
+
+A sampler would consume the same arguments it does today:
+
+- span name
+- span context
+- span kind
+- span attributes
+- span links
+
+The sampler would be called twice: once at creation, and once at span end.
+
+The span's sampling decision would be determined by the intersection of the
+sampling decision made by either ShouldSample call:
+
+| First ShouldSample result | Second ShouldSample result | Final Result       |
+|---------------------------|----------------------------|--------------------|
+| NOT_RECORD                | N/A (Don't Call)           | NOT_RECORD         |
+| RECORD_AND_SAMPLED        | N/A (Don't Call)           | RECORD_AND_SAMPLED |
+| RECORD                    | NOT_RECORD                 | NOT_RECORD         |
+| RECORD                    | RECORD                     | RECORD             |
+| RECORD                    | RECORD_AND_SAMPLED         | RECORD_AND_SAMPLED |
+
+## Internal details
+
+### Idempotence of samplers
+
+The calling of shouldSample multiple times can result in inconsistent results
+if the sampler is truly random. As such, the samplers will need to be idempotent
+based on the input passed. For example, a probability sampler would need to
+use data present in the passed parameters, such as the span id, to return sampling
+decisions.
```

Comment:
not necessarily. In Jaeger, the [sampling decision][] explicitly contains a `Retryable` flag. A probabilistic sampler would return `retryable=false`, because it does not depend on the content of the span and therefore there's no point calling it more than once (in fact it's detrimental since it will skew the weight of the probability, which breaks inference of other statistics). We also support a [PrioritySampler][] that can combine multiple simpler samplers and short-circuit their execution once all retryable decisions have been exhausted. In practice this allows efficient sampling in the simplest cases, while also supporting customized sampling policies.

[sampling decision]: https://github.com/jaegertracing/jaeger-client-go/blob/master/sampler_v2.go#L18
[PrioritySampler]: https://github.com/jaegertracing/jaeger-client-go/blob/f2bdef7faf35d024087d9225d88a202a86e21a31/x/priority_sampler.go#L66

Reply:
I'll amend this section with that option.

Reply:
done.

Reply:
I was doing some thinking of this concept, and I'm debating adding this to the OTEP, as it might be required to ensure optimal performance.

The clear use case is for a sampler that is truly random, doesn't depend on the contents of the span, and makes the decision once and doesn't need to make it again.

The second use case is when there is some level of complex logic to determine a span decision. For example, a sampler that depends on various http route attributes to rate limit them (e.g. http path + method + status code). Assuming that the final interface does provide more hooks for samplers to update their decision (e.g. on every attribute update), then it would be valuable to notify the span when no further shouldSample calls should be made. 

This could be deferred to a further OTEP, but to your point this does allow more complex chaining of samplers as you described with PrioritySampler.